p2panda-rs/src/storage_provider/traits/entry_store.rs

```diff
@@ -0,0 +1,295 @@
+use async_trait::async_trait;
+use bamboo_rs_core_ed25519_yasmf::entry::is_lipmaa_required;
+
+use crate::entry::LogId;
+use crate::entry::SeqNum;
+use crate::hash::Hash;
+use crate::identity::Author;
+use crate::schema::SchemaId;
+use crate::storage_provider::errors::EntryStorageError;
+use crate::storage_provider::traits::AsStorageEntry;
+
+/// Trait which handles all storage actions relating to `Entry`.
+///
+/// This trait should be implemented on the root storage provider struct. It's definitions make up
+/// the required methods for inserting and querying entries from storage.
+#[async_trait]
+pub trait EntryStore<StorageEntry: AsStorageEntry> {
+    /// Insert an entry into storage.
+    async fn insert_entry(&self, value: StorageEntry) -> Result<bool, EntryStorageError>;
+
+    /// Returns entry at sequence position within an author's log.
+    async fn entry_at_seq_num(
+        &self,
+        author: &Author,
+        log_id: &LogId,
+        seq_num: &SeqNum,
+    ) -> Result<Option<StorageEntry>, EntryStorageError>;
+
+    /// Returns the latest Bamboo entry of an author's log.
+    async fn latest_entry(
+        &self,
+        author: &Author,
+        log_id: &LogId,
+    ) -> Result<Option<StorageEntry>, EntryStorageError>;
+
+    /// Return vector of all entries of a given schema.
+    async fn by_schema(&self, schema: &SchemaId) -> Result<Vec<StorageEntry>, EntryStorageError>;
```

Comment:
This could also be merged together with other query methods to one large `query(filter: EntryQuery)` (or similar) function where:

```rust
struct EntryQuery {
  schemas: Option<Vec<DocumentViewId>>,
  seq_nums: Option<Vec<SeqNum>>,
  authors: ...
  ...
}
```

I don't think it's necessary, just wanted to drop that idea here as it's reminding me of Aljoschas "ReqRes".

Reply:
Upside with that is that it could be easier to introduce pagination like this one day for future queries like browsing all schemas etc.

Reply:
Yeh, this would be great! There are a few things that can be raised as issues for after this PR is merged. I'll add this to the list.